File: backend/app/routers/prelabel.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.config import get_settings
from app.db.session import get_db
from app.schemas.prelabel import (
    LabelStudioStatusSyncResponse,
    PredictionPreviewResponse,
    PrelabelJobStatusResponse,
    PrelabelRunResponse,
    TaskImageStatusResponse,
)
from app.services.prelabel_service import PrelabelService

router = APIRouter(prefix="/api", tags=["prelabel"])


@router.post("/tasks/{task_id}/prelabel", response_model=PrelabelRunResponse)
def run_prelabel(
    task_id: str,
    db: Session = Depends(get_db),
) -> PrelabelRunResponse:
    service = PrelabelService(get_settings())
    return service.run_prelabel(db=db, task_id=task_id)


@router.get("/prelabel-jobs/{job_id}", response_model=PrelabelJobStatusResponse)
def get_prelabel_job(
    job_id: str,
    db: Session = Depends(get_db),
) -> PrelabelJobStatusResponse:
    service = PrelabelService(get_settings())
    return service.get_prelabel_job(db=db, job_id=job_id)


@router.get("/tasks/{task_id}/images", response_model=TaskImageStatusResponse)
def list_task_images(
    task_id: str,
    db: Session = Depends(get_db),
) -> TaskImageStatusResponse:
    service = PrelabelService(get_settings())
    return service.list_task_images(db=db, task_id=task_id)


@router.post("/tasks/{task_id}/sync-label-studio-status", response_model=LabelStudioStatusSyncResponse)
def sync_label_studio_status(
    task_id: str,
    db: Session = Depends(get_db),
) -> LabelStudioStatusSyncResponse:
    service = PrelabelService(get_settings())
    return service.sync_label_studio_status(db=db, task_id=task_id)


@router.get("/tasks/{task_id}/images/{image_id}/prediction-preview", response_model=PredictionPreviewResponse)
def get_prediction_preview(
    task_id: str,
    image_id: str,
    db: Session = Depends(get_db),
) -> PredictionPreviewResponse:
    service = PrelabelService(get_settings())
    return service.get_prediction_preview(db=db, task_id=task_id, image_id=image_id)

File: frontend/src/api/prelabel.ts
import type {
  LabelStudioStatusSyncResponse,
  PredictionPreviewResponse,
  PrelabelRunRequest,
  PrelabelJobStatusResponse,
  PrelabelRunResponse,
  TaskImagesResponse,
} from "../types/api";
import { apiClient } from "./client";

export async function runPrelabel(taskId: string, payload?: PrelabelRunRequest) {
  const response = await apiClient.post<PrelabelRunResponse>(`/api/tasks/${taskId}/prelabel`, payload ?? {});
  return response.data;
}

export async function getPrelabelJob(jobId: string) {
  const response = await apiClient.get<PrelabelJobStatusResponse>(`/api/prelabel-jobs/${jobId}`);
  return response.data;
}

export async function getTaskImages(taskId: string) {
  const response = await apiClient.get<TaskImagesResponse>(`/api/tasks/${taskId}/images`);
  return response.data;
}

export async function syncLabelStudioStatus(taskId: string) {
  const response = await apiClient.post<LabelStudioStatusSyncResponse>(`/api/tasks/${taskId}/sync-label-studio-status`);
  return response.data;
}

export async function getPredictionPreview(taskId: string, imageId: string) {
  const response = await apiClient.get<PredictionPreviewResponse>(
    `/api/tasks/${taskId}/images/${imageId}/prediction-preview`,
  );
  return response.data;
}
